cmd/influxd/export/export.go

```diff
@@ -0,0 +1,367 @@
+		Database:        cmd.database,
+		RetentionPolicy: retentionPolicy,
+		ShardID:         id,
+		ExportStart:     tsStart,
+		ExportEnd:       tsEnd,
+	}
+
+	// TODO: verify shard backup data
+	return cmd.downloadAndVerify(req, shardArchivePath, nil)
+}
+
+// exportDatabase will request the database information from the server and then backup the metastore and
+// every shard in every retention policy in the database. Each shard will be written to a separate tar.
+func (cmd *Command) exportDatabase(tsStart, tsEnd time.Time) error {
+	cmd.StdoutLogger.Printf("backing up db=%s start %s end %s", cmd.database, tsStart, tsEnd)
+
+	req := &snapshotter.Request{
+		Type:     snapshotter.RequestDatabaseInfo,
+		Database: cmd.database,
+	}
+	cmd.StdoutLogger.Println("made it to before the export response paths")
+	response, err := cmd.requestInfo(req)
+	if err != nil {
+		return err
+	}
+	cmd.StdoutLogger.Println("made it to the export response paths")
+	return cmd.exportResponsePaths(response, tsStart, tsEnd)
+}
+
+// exportRetentionPolicy will request the retention policy information from the server and then backup
+// the metastore and every shard in the retention policy. Each shard will be written to a separate tar.
+func (cmd *Command) exportRetentionPolicy(retentionPolicy string, tsStart, tsEnd time.Time) error {
+	cmd.StdoutLogger.Printf("backing up rp=%s start %s end %s", retentionPolicy, tsStart, tsEnd)
+
+	req := &snapshotter.Request{
+		Type:            snapshotter.RequestRetentionPolicyInfo,
+		Database:        cmd.database,
+		RetentionPolicy: retentionPolicy,
+	}
+
+	response, err := cmd.requestInfo(req)
+	if err != nil {
+		return err
+	}
+
+	return cmd.exportResponsePaths(response, tsStart, tsEnd)
+}
+
+// exportResponsePaths will backup the metastore and all shard paths in the response struct
+func (cmd *Command) exportResponsePaths(response *snapshotter.Response, tsStart, tsEnd time.Time) error {
+	if err := cmd.exportMetastore(); err != nil {
+		return err
+	}
+
+	// loop through the returned paths and back up each shard
+	for _, path := range response.Paths {
+		rp, id, err := retentionAndShardFromPath(path)
+		if err != nil {
+			return err
+		}
+
+		if err := cmd.exportShard(rp, id, tsStart, tsEnd); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// exportMetastore will backup the metastore on the host to the passed in path. Database and retention policy backups
+// will force a backup of the metastore as well as requesting a specific shard backup from the command line
+func (cmd *Command) exportMetastore() error {
+	metastoreArchivePath, err := cmd.nextPath(filepath.Join(cmd.path, Metafile))
+	if err != nil {
+		return err
+	}
+
+	cmd.StdoutLogger.Printf("backing up metastore to %s", metastoreArchivePath)
+
+	req := &snapshotter.Request{
+		Type: snapshotter.RequestMetastoreBackup,
+	}
+
+	return cmd.downloadAndVerify(req, metastoreArchivePath, func(file string) error {
+		binData, err := ioutil.ReadFile(file)
```

Comment:
Reading in an entire file just to check 8 bytes? Seems silly. I would suggest either verifying the entire file (can decode the blob) or just reading in the magic bytes.

Reply:
After looking further, I think this blob should definitely just be decoded and used as the source of metadata for which shards to pull down, and not rely on shard paths that don't contain timestamp info.

Reply:
optimized this to get just the first 8 bytes.  See above RE: metadata crawling.  probably I will get to this, since we'll be doing work on the OSS import/restore side next.